Support for openfl.filters.BevelFilter

scanline opened the pull request:
At the moment the BevelFilter is completely missing - not even a stub. A quick example like this:
```
package;

import openfl.display.Shape;
import openfl.display.Sprite;
import openfl.filters.BevelFilter;
import openfl.text.TextField;
import openfl.text.TextFormat;

class Main extends Sprite
{

	public function new()
	{
		super();
		var button:Sprite = createButton(10, 10);
		addChild(button);

		button = createButton(310, 10);
		button.filters = [new BevelFilter(14, 45, 0xFFFFFF, 0.5, 0x000000, 0.5, 4, 4, 1, 1, "inner", false)];
		addChild(button);

		button = createButton(10, 120);
		button.filters = [new BevelFilter(14, 45, 0xFFFFFF, 0.5, 0x000000, 0.5, 4, 4, 1, 1, "inner", true)];
		addChild(button);

		button = createButton(310, 120);
		button.filters = [new BevelFilter(14, 45, 0xFFFFFF, 0.5, 0x000000, 0.5, 4, 4, 1, 1, "outer", false)];
		addChild(button);

		button = createButton(10, 230);
		button.filters = [new BevelFilter(14, 45, 0xFFFFFF, 0.5, 0x000000, 0.5, 4, 4, 1, 1, "full", false)];
		addChild(button);
	}

	private function createButton(x:Float, y:Float):Sprite
	{
		var sprite:Sprite = new Sprite();

		var shape = new Shape();
		shape.graphics.beginFill(0xd3d3d3, 1);
		shape.graphics.drawRoundRect(0, 0, 280, 96, 20, 20);
		shape.graphics.endFill();
		sprite.addChild(shape);

		var format:TextFormat = new TextFormat();
		format.size = 48;
		format.font = "_sans";
		format.align = "center";

		var tField:TextField = new TextField();
		tField.autoSize = "left";
		tField.defaultTextFormat = format;
		tField.text = "Button";
		tField.y = 48 - tField.height / 2;
		tField.x = 140 - tField.width / 2;
		sprite.addChild(tField);
		sprite.x = x;
		sprite.y = y;
		return sprite;
	}
}
```

which the Flashplayer displays like this:

![BevelFilterFlash](https://github.com/openfl/openfl/assets/22029272/ca6e4e89-4872-4897-9c13-591c26a60b61)

just throws **Type not found : openfl.filters.BevelFilter** when targeting e.g. html5.

With this pull request OpenFL now supports the complete features of the BevelFilter and things will look like this:
![BevelFilter](https://github.com/openfl/openfl/assets/22029272/f0a4e41e-782e-44d8-841e-93f2aa04c96a)

`BitmapFilterType.INNER` with `knockout` set to true and false work flawlessly ~while `BitmapFilterType.OUTER` and `BitmapFilterType.FULL` suffer from the same fate as any other of OpenFL's OpenGL-based filters (GlowFilter and the like):
OpenFL does not allow dynamic resizing of textures to make room for the filter's effect~.

joshtynjala committed:
Graphics: fix readGraphicsData() returning incorrect caps/joints in line style (closes #2659)

The null default values from the lineStyle() function signature were propagating and being converted to 0, which changed the caps/joints style when running drawGraphicsData() with the result of readGraphicsData()

dimensionscape committed:
Incl option to skip text cache

Since the measurement cache can be problematic in some use cases, I decided to include the `skip_measurement_cache` flag in the internals. This define is likely to become obsolete and removed in future versions.

dimensionscape committed:
TextEngine: Remove extra return

Woops

scanline committed:
Create BevelFilter.hx

Implements openfl.filters.BevelFilter

joshtynjala commented:
I switched the base branch to 9.4.0-Dev since the 9.3.0 release is happening soon, and this should have more time for testing once merged.

scanline committed:
Update BevelFilter.hx

Taking care of the extra area caused by applying this filter to a DisplayObject

scanline commented:
As it turns out I was wrong. After digging through openfl.display.DisplayObjectRenderer.hx
https://github.com/openfl/openfl/blob/bc334e20dc87def193fa52b22acca265d5e749b5/src/openfl/display/DisplayObjectRenderer.hx#L371

I realized that OpenFL **does** take care of the extra area caused by a filter now. We just need to tell it the new size.
Well, with the latest commit the outcome of this filter will now look close to the Flashplayer even for `BitmapFilterType.FULL` and `BitmapFilterType.OUTER`.
![BevelFilterUpdated](https://github.com/openfl/openfl/assets/22029272/5beffa43-9c4c-447d-991a-411de2e4a8b5)

joshtynjala commented:
This works correctly in Chrome and Safari. However, in Firefox, there's an exception thrown. If you can fix it, I'll be happy to merge!

> [openfl.display.Shader] ERROR: Unable to initialize the shader program
> Uniform `uBevelType` is not linkable between attached shaders.

Shaunmax commented:
> This works correctly in Chrome and Safari. However, in Firefox, there's an exception thrown. If you can fix it, I'll be happy to merge!
> 
> > [openfl.display.Shader] ERROR: Unable to initialize the shader program
> > Uniform `uBevelType` is not linkable between attached shaders.

@scanline were you able to fix the issue happening on Firefox?

scanline commented:
Sorry, I didn't notice that there was a problem. I've solved the issue with FireFox now. As I did not know how to reopen my existing pull request, I created a new one. #2759

joshtynjala commented:
> As I did not know how to reopen my existing pull request

I think if you push new commits to the existing branch in your fork, it should notify us of changes to the PR. Additional comments will also send us notifications, and we can reopen manually if needed.

For future PRs. The current duplicate PR is fine, and we can leave this one closed.

scanline commented:
Copy that Josh!
